main.cc

```diff
@@ -1444,14 +1450,32 @@ run()
                  * don't really care about resizing, because a tiling
                  * WM isn't going to give us what we asked for anyway!
                  */
-                if (e.window.event == SDL_WINDOWEVENT_RESIZED)
+                switch (e.window.event) {
+                case SDL_WINDOWEVENT_RESIZED:
                     resize(e.window.data1, e.window.data2);
-                break;
+                    break;
+
+                case SDL_WINDOWEVENT_FOCUS_LOST:
+                    wnd.has_focus = false;
+                    break;
+
+                case SDL_WINDOWEVENT_FOCUS_GAINED:
+                    wnd.has_focus = true;
+                    break;
+                }
 
             case SDL_MOUSEMOTION:
-                mouse_axes[EN_MOUSE_AXIS(input_mouse_x)] += e.motion.xrel;
-                mouse_axes[EN_MOUSE_AXIS(input_mouse_y)] += e.motion.yrel;
+            {
+                auto x = e.motion.xrel;
+                auto y = e.motion.yrel;
+
+                x = x > 0 ? std::min(MAX_AXIS_PER_EVENT, x) : std::max(-MAX_AXIS_PER_EVENT, x);
```

Comment:
(It would be ideal to write this as:

`auto x = clamp(e.motion.xrel, -MAX_AXIS_PER_EVENT, MAX_AXIS_PER_EVENT);`